dask/utils.py

```diff
@@ -620,17 +620,33 @@ def wrapped(*args, **kwargs):
     return wrapper
 
 
-def funcname(func, full=False):
+def funcname(func):
     """Get the name of a function."""
-    while hasattr(func, 'func'):
-        func = func.func
+    # functools.partial
+    if isinstance(func, functools.partial):
+        return funcname(func.func)
+    # methodcaller
+    if isinstance(func, methodcaller):
+        return func.method
+
+    module_name = getattr(func, '__module__', None) or ''
+    type_name = getattr(type(func), '__name__', None) or ''
+
+    # toolz.curry
+    if 'toolz' in module_name and 'curry' == type_name:
+        return func.func_name
+    # multipledispatch objects
+    if 'multipledispatch' in module_name and 'Dispatcher' == type_name:
+        return func.name
+
+    # All other callables
     try:
-        if full:
-            return func.__qualname__.strip('<>')
```

Comment:
This function is only used to provide names for keys in the graph, and annotations in the diagnostics. Since `__qualname__` is only available in Py3, and the option was never used, I'd rather avoid the (slight) increase in complexity. Not a strong opinion though.